### PR TITLE
Improve `wsinterp` docstrings and test function

### PR DESCRIPTION
**Pull request by bobleesj**

### CODE REVIEW

File: src/diffpy/utils/resampler.py [hunk 1]
```diff
@@ -15,52 +15,50 @@
 
 """Various utilities related to data parsing and manipulation."""
 
-import numpy
```
Comment (bobleesj) [inline src/diffpy/utils/resampler.py:18]: import `numpy` as `np` for consistency

File: src/diffpy/utils/__init__.py [hunk 2]
```diff
@@ -20,5 +20,3 @@
 
 # silence the pyflakes syntax checker
 assert __version__ or True
-
-# End of file
```
Comment (bobleesj) [inline src/diffpy/utils/__init__.py:24]: Remove `# End of file` as no longer needed. Unless I am not sure it has a specific usecase.

File: src/diffpy/utils/resampler.py [hunk 3]
```diff
@@ -15,52 +15,50 @@
...
+import numpy as np
 
 
-# NOTE - this should be faster than resample below and conforms more closely to
-# numpy.interp. I'm keeping resample for legacy reasons.
 def wsinterp(x, xp, fp, left=None, right=None):
```
Comment (bobleesj) [inline src/diffpy/utils/resampler.py:21]: I was reading the code again, but I thought `wsinterp` docstrings could be improved by making it explicit.

**Review by bobleesj — COMMENTED** [review]
Ready for review - pls see my in-line comments.

File: src/diffpy/utils/resampler.py [hunk 4]
```diff
@@ -15,25 +15,24 @@
...
     """One-dimensional Whittaker-Shannon interpolation.
 
-    This uses the Whittaker-Shannon interpolation formula to interpolate the value of fp (array),
-    which is defined over xp (array), at x (array or float).
+    Reconstruct a continuous signal from discrete data points by utilizing sinc functions
```
Comment (sbillinge) [inline src/diffpy/utils/resampler.py:24]: We need to keep a reference to Whitaker-Shannon here, preferably a reference to their paper, though I very much like your expanded description so let's keep that.

File: src/diffpy/utils/resampler.py [hunk 5]
```diff
@@ -15,25 +15,24 @@
...
     ----------
     x: ndarray
-        Desired range for interpolation.
+        The x values at which interpolation is computed.
     xp: ndarray
-        Defined range for fp.
+        The array of known x values.
     fp: ndarray
-        Function to be interpolated.
+        The array of y values associated xp.
```
Comment (sbillinge) [inline src/diffpy/utils/resampler.py:35]: These are MUCH better, thank you!

File: src/diffpy/utils/resampler.py [hunk 6]
```diff
@@ -43,24 +42,23 @@ def wsinterp(x, xp, fp, left=None, right=None):
 
     Returns
     -------
-    float:
-        If input x is a scalar (not an array), return the interpolated value at x.
-    ndarray:
-        If input x is an array, return the interpolated array with dimensions of x.
+    ndarray or float
+        Interpolated values at points x. Returns a single float if x is a scalar,
```
Comment (sbillinge) [inline src/diffpy/utils/resampler.py:46]: Always start with "The"

File: tests/test_resample.py [hunk 7]
```diff
@@ -1,26 +1,34 @@
...
-    # Interpolate a few random datasets
+    # Generating fp values across 10 trial runs
     trials = 10
-    for trial in range(trials):
-        fp = np.array([random.random() * ssr for i in range(-n, n + 1, 1)])
+
+    for _ in range(trials):
+        # Create random function values (fp) at the points defined in xp above
+        fp = np.array([random.random() * ssr for _ in range(-n, n + 1, 1)])
+
```
Comment (sbillinge) [inline tests/test_resample.py:27]: Let's close this up. No need for the blank lines that are not adding subtracting value.  This is true elsewhere in this file too

File: src/diffpy/utils/resampler.py (hunk 6 above)
Comment (bobleesj) [reply src/diffpy/utils/resampler.py:46]: done

File: src/diffpy/utils/resampler.py (hunk 5 above)
Comment (bobleesj) [reply src/diffpy/utils/resampler.py:35]: 👍

File: src/diffpy/utils/resampler.py (hunk 4 above)
Comment (bobleesj) [reply src/diffpy/utils/resampler.py:24]: Done - DOI added.

**Review by bobleesj — COMMENTED** [review]
@sbillinge ready for review! - added DOI, didn't add any additional empty liens

**Comment by bobleesj** [comment]
> @sbillinge ready for review! - added DOI, didn't add any additional empty liens

Screenshot 2024-12-13 at 10 29 53 AM

@sbillinge it passes - just codecov..

**Comment by sbillinge** [comment]
This looks great, o can merge when it passes. 

Nor I am greedy though..... Would it be nice if it could take, optionally, a DiffractionObject and return a distraction object?  Or maybe it is better to make a method on the DO itself?

**Comment by bobleesj** [comment]
Screenshot 2024-12-13 at 10 31 06 AM

@sbillinge `v5` has been released. Shall we migrate from our current `v4` to `v5` for Codecov actions?

**Comment by bobleesj** [comment]
> Nor I am greedy though..... Would it be nice if it could take, optionally, a DiffractionObject and return a distraction object? Or maybe it is better to make a method on the DO itself?

having `wsinterp` as a modular and standalone function,  and then have `DiffractionObject` integrate `wsinterp` seems like a scalable approach to me imo?

**Comment by sbillinge** [comment]
> > Nor I am greedy though..... Would it be nice if it could take, optionally, a DiffractionObject and return a distraction object? Or maybe it is better to make a method on the DO itself?
> 
> having `wsinterp` as a modular and standalone function, and then have `DiffractionObject` integrate `wsinterp` seems like a scalable approach to me imo?

Yah my thinking too. Actually it is for pdfs, not for diffraction, but for sure there is a PDFObject in our future.

**Comment by bobleesj** [comment]
> Yah my thinking too. Actually it is for pdfs, not for diffraction, but for sure there is a PDFObject in our future.

I left a comment in the PDF object issue here: https://github.com/diffpy/diffpy.utils/issues/47

Maybe we could merge this, once https://github.com/Billingegroup/release-scripts/pull/109 is merged all the way to `v0`.

**Pull request merged**